File: cpp/LeetCode/No52.cpp
/*
The n-queens puzzle is the problem of placing n queens on an n x n chessboard such that no two queens attack each other.

Given an integer n, return the number of distinct solutions to the n-queens puzzle.
*/
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int totalNQueens(int n) {
        size = n;
        result = 0;
        vector<bool> col = vector<bool>(n);
        vector<bool> dia = vector<bool>(2*n-1);
        vector<bool> adia = vector<bool>(2*n-1);
        backtrack(col, dia, adia);
        return result;
    }
private:
    int size = 0;
    int result = 0;
    void backtrack(vector<bool> &col, vector<bool> &dia, vector<bool> &adia, int row=0){
        if(row==size){
            result++;
            return;
        }
        for(int i=0;i<size;i++){
            if(col[i]){
                continue;
            }
            if(dia[row-i+size-1]){
                continue;
            }
            if(adia[row+i]){
                continue;
            }
            col[i] = true;
            dia[row-i+size-1] = true;
            adia[row+i] = true;
            backtrack(col, dia, adia, row+1);
            col[i] = false;
            dia[row-i+size-1] = false;
            adia[row+i] = false;
        }
    }
};

int main(void){
    Solution solution;
    int n = 4;
    cout << solution.totalNQueens(n) << endl;

    return 0;
}

File: python/LeetCode Quiz/No52.py
'''
The n-queens puzzle is the problem of placing n queens on an n x n chessboard such that no two queens attack each other.

Given an integer n, return the number of distinct solutions to the n-queens puzzle.
'''
class Solution:
    def totalNQueens(self, n: int) -> int:
        self.__res = 0
        def backtrack(row: int = 0, col: set = set(), dia: set = set(), adia: set = set()):
            if row==n:
                self.__res+=1
                return
            for i in range(n):
                if i in col:
                    continue
                if row-i in dia:
                    continue
                if row+i in adia:
                    continue
                col.add(i)
                dia.add(row-i)
                adia.add(row+i)
                backtrack(row+1, col, dia, adia)
                col.remove(i)
                dia.remove(row-i)
                adia.remove(row+i)
        backtrack()
        return self.__res

def main():
    solution = Solution()

    n = 1
    print(solution.totalNQueens(n))

if __name__=='__main__':
    main()
